### ISSUE TITLE
[MQTT] Reestablish subscriptions after reconnect

### ISSUE THREAD

**Issue by ffendt** [akka/alpakka#1824]
Hi there,

I am a developer of Eclipse Ditto which uses akka-stream-alpakka-mqtt and think I found a little bug concerning its reconnecting functionality.

The issue seems to be that after a reconnect (because e.g. the broker was unreachable) the subscriptions aren't reestablished. So after a reconnect I am not able to receive messages any longer (publishing will work though).

When debugging my application, I can see that [MqttFlowStage#connectComplete](https://github.com/akka/alpakka/blob/master/mqtt/src/main/scala/akka/stream/alpakka/mqtt/impl/MqttFlowStage.scala#L205) callback is called by the paho client after a successful reconnect.
The paho team suggests [to resubscribe](https://github.com/eclipse/paho.mqtt.c/issues/221#issuecomment-276784143) in the callback. As far as I understand (with my very little Scala knowledge) the subscriptions are only established once in [MqttFlowStage#onConnect](https://github.com/akka/alpakka/blob/master/mqtt/src/main/scala/akka/stream/alpakka/mqtt/impl/MqttFlowStage.scala#L104) which is called after the initial connect (see [MqttFlowStage#preStart](https://github.com/akka/alpakka/blob/master/mqtt/src/main/scala/akka/stream/alpakka/mqtt/impl/MqttFlowStage.scala#L275) ).

In the gitter chat, @2m suggested to move the subscription logic to `connectComplete`. As my company has relatively strict rules on contributing code, I don't know yet if or when I will be allowed to provide a PR for this.

Best regards
Florian

**Comment by 2m** [comment]
Thanks for the ticket. As a workaround one could try disabling Paho reconnect logic and instead use [`RestartSource`](https://doc.akka.io/docs/akka/current/stream/operators/RestartSource/onFailuresWithBackoff.html) from akka-stream.

Another workaround is to use [Alpakka Mqtt Streaming connector](https://doc.akka.io/docs/alpakka/current/mqtt-streaming.html), however that connector provides quite a low level API (different API is being worked on in https://github.com/akka/alpakka/pull/1565)

**Comment by huntc** [comment]
Should you have to re-subscribe given the continuation of a session? I do t recall the spec stating that you should need to as the previous subscriptions should be continued.

**Comment by 2m** [comment]
@ffendt have you tried `MqttConnectionSettings.withCleanSession(session)`?

**Comment by ffendt** [comment]
The mqtt streaming connector currently is no real option for us, since as you said it only provides a very low level API which we don't want to use as of now.

I also stumbled upon the `cleanSession` setting which I also tried out and failed. But while writing I think that might have been caused by how I'm reproducing the problem. Currently I am testing against [docker.io/eclipse-mosquitto:1.4.12](https://hub.docker.com/_/eclipse-mosquitto) by simply stopping and starting the container. But that will probably destroy the session on the broker, which would be a reason why the subscriptions aren't available after a reconnect. I'll try to find a way to test this without killing the broker. Thanks for now, i'll get back to you as soon as I know more.
